### ISSUE TITLE
[Probably a bug] History API: gaps in weather_code soon after midnight

### ISSUE THREAD

**Issue by PavelSyomin** [open-meteo/open-meteo#1028]
Hi,

While using Open-Meteo history API in production, our team has faced a strange issue occurring irregularly. The workflow is as follows:
- we have ~1000 “weather stations” in our system, that is, a geographical point with name, roughly an approximation of a particular city;
- our weather processing service queries History API every day soon after UTC midnight to get and store weather history for the preceding day;
- before storing the response, it checks if the received data is consistent. In particular, it checks if there are any gaps in data for any variable, that is, `None` value inside a row of non-null values, e. g. `0 1 2 None 4 5`, warns about any gaps found, and fails to store the data if there are gaps in any of the weather variables;
- it is expected that there are no such gaps in the data. However, every day a number of weather stations ranging from 4 to more than 600 fails to load history because of gaps in `weather_code` variable;
- the weather processing service is designed to re-attempt history loading on failure, so it retries to obtain data for the same weather stations again a hour after the initial attempt. This second attempt is always successful, that is, there are no gaps in `weather_code` variable that have existed an hour before;
- as a result, the issue is not severe but annoying because of large number of warnings in logs, and probably it also related to some bug in the API that is better to be fixed.

In short, the problem is a gap in `weather_code` variable (e.g. `0 0 None 1 0 81`) that appears in the response of History API for some specific coordinates in the beginning of the next day and typically exists for an hour or slightly longer.

Unfortunately, right now I cannot attach the samples of incorrect data, because out weather processing service does not log the entire data that have caused the problem. If the specific requests and responses are required to reproduce the issue, I can try to modify the logging and provide for the necessary details.

I hope that this report will help you to improve Open-Meteo and solve problems (if they do exist).

With regards,

Pavel

**Comment by patrick-zippenfenig** [comment]
Hi,
it would be great to have an example URL and returned data. Currently, I am not sure what could cause the issue. Data is always processed as a time-series without any gaps in the middle.

**Comment by PavelSyomin** [comment]
Hi,

One of the URLs with the problem was https://archive-api.open-meteo.com/v1/archive?latitude=50.267&longitude=127.533&start_date=2024-09-19&end_date=2024-09-19&hourly=temperature_2m,relative_humidity_2m,rain,weather_code,surface_pressure,cloud_cover,wind_speed_10m,wind_direction_10m,soil_temperature_0_to_7cm,soil_moisture_0_to_7cm&wind_speed_unit=ms&models=ecmwf_ifs The record in our log has a timestamp `Sep 21, 2024 @ 06:21:19.302` and a message about the gap in `weather_code` variable. Similar failures (the same problem for the same URL) were also logged at `Sep 19, 2024 @ 06:29:00.189`, `Sep 18, 2024 @ 06:28:54.353`, and `Sep 6, 2024 @ 06:28:43.361`.

Unfortunately, I still cannot attach the response data obtained from the Open-Meteo API at these timestamps because the necessary changes in the logging have not been made yet. But probably the sample URLs will help you investigate this issue. If they don't, please write about it here, and I will try to provide exhaustive logging details (URL, timestamp, returned data).

**Comment by patrick-zippenfenig** [comment]
An error in this dataset is rather unlikely because it is updating 24 hours at a time and gaps in the middle should not be possible. Could it be possible that the processing on your side is causing the issues?

**Comment by PavelSyomin** [comment]
Seems that the problem is with data. We've added the logging of Open-Meteo response that had caused the problem, and here it is.

URL: https://archive-api.open-meteo.com/v1/archive?latitude=43.799&longitude=131.951&start_date=2024-09-24&end_date=2024-09-24&hourly=temperature_2m,relative_humidity_2m,rain,weather_code,surface_pressure,cloud_cover,wind_speed_10m,wind_direction_10m,soil_temperature_0_to_7cm,soil_moisture_0_to_7cm&wind_speed_unit=ms&models=ecmwf_ifs
Request timestamp: Sep 26, 2024 @ 06:21:13.873 UTC
Response: [resp.json](https://github.com/user-attachments/files/17144899/resp.json)

There are data for some variables (e. g. `temperature_2m`), but almost no data for other variables (e. g. `relative_humidity_2m`), as well as previously reported gaps in `weather_code`.

**Comment by PavelSyomin** [comment]
Hi,

Are there any updates regarding this issue?

If additional information from my part is necessary, please let me know.

**Comment by patrick-zippenfenig** [comment]
I deployed a fix making sure that weather codes are only calculated if all data is available. It looks like data was requested, exactly at the time of update. In this case precipitation data was already available leading to code `51` while cloud cover was not yet available. The fix makes sure that all data needs to be present before weather codes are calculated

**Comment by PavelSyomin** [comment]
Seems that it have worked: no warnings about the gaps have been observed for 2 weeks. Thank you for solving this problem!
